include/pqxx/blob.hxx

```diff
@@ -0,0 +1,207 @@
+/* Binary Large Objects interface.
+ *
+ * Read or write large objects, stored in their own storage on the server.
+ *
+ * DO NOT INCLUDE THIS FILE DIRECTLY; include pqxx/largeobject instead.
+ *
+ * Copyright (c) 2000-2020, Jeroen T. Vermeulen.
+ *
+ * See COPYING for copyright license.  If you did not receive a file called
+ * COPYING with this source code, please notify the distributor of this
+ * mistake, or contact the author.
+ */
+#ifndef PQXX_H_BLOB
+#define PQXX_H_BLOB
+
+#include "pqxx/compiler-public.hxx"
+#include "pqxx/internal/compiler-internal-pre.hxx"
+
+#include <cstdint>
+
+#include "pqxx/dbtransaction.hxx"
+
+
+namespace pqxx
+{
+/** Binary large object.
+ *
+ * This is how you store data that may be too large for the @c BYTEA type.
+ * Access operations are similar to those for a file: you can read, write,
+ * query or set the current reading/writing position, and so on.
+ *
+ * These large objects live in their own storage on the server, indexed by an
+ * integer object identifier ("oid").
+ *
+ * Two @c blob objects may refer to the same actual large object in the
+ * database at the same time.  Each will have its own reading/writing position,
+ * but writes to the one will of course affect what the other sees.
+ */
+class PQXX_LIBEXPORT blob
+{
+public:
+  /// Create a new, empty large object.
+  /** You may optionally specify an oid for the new blob.  If you do, then
+   * the new object will have that oid -- or creation will fail if there
+   * already is an object with that oid.
+   */
+  [[nodiscard]] static oid create(dbtransaction &, oid = 0);
+
+  /// Delete a large object, or fail if it does not exist.
+  static void remove(dbtransaction &, oid);
+
+  /// Open blob for reading.  Any attempt to write to it will fail.
+  [[nodiscard]] static blob open_r(dbtransaction &, oid);
+  // Open blob for writing.  Any attempt to read from it will fail.
+  [[nodiscard]] static blob open_w(dbtransaction &, oid);
+  // Open blob for reading and/or writing.
+  [[nodiscard]] static blob open_rw(dbtransaction &, oid);
+
+  /// You can default-construct a blob, but it won't do anything useful.
+  /** Most operations on a default-constructed blob will throw @c usage_error.
+   */
+  blob() = default;
+
+  /// You can move a blob, but not copy it.  The original is rendered unusable.
+  blob(blob &&);
+  /// You can move a blob, but not copy it.  The original is rendered unusable.
+  blob &operator=(blob &&);
+
+  blob(blob const &) = delete;
+  blob &operator=(blob const &) = delete;
+  ~blob();
+
+  /// Maximum number of bytes that can be read or written at a time.
+  /** The underlying protocol only supports reads and writes up to 2 GB
+   * exclusive.
+   *
+   * If you need to read or write more data to or from a binary large object,
+   * you'll have to break it up into chunks.
+   */
+  static constexpr std::size_t chunk_limit = 0x7fffffff;
```

Comment:
Use INT32_MAX or some constant like `std::numeric_limits<std::size_t>::max()` or  `std::numeric_limits<int32_t>::max()`?

Reply:
I struggled with that question as well.  The PostgreSQL documentation interchangeably says `INT_MAX` and "2 GB," so that's not much help.  But based on what it says, if you had a client where `int` was 64 bits wide, it is not clear to me that that limit would change.

I could look at what the code does, but once we start seeing such systems, the postgres team might decide that the code was incorrect in a way that previously had no consequences.

Things may become clearer in the future, when an actual problem starts showing up.  Until then, I think it's better to apply (and document) the strictest version of the limit.